File: python/500-KeyboardRow/main.py
from typing import List
import unittest

class Solution:
    def findWords(self, words: List[str]) -> List[str]:
        top = set("qwertyuiop")
        middle = set("asdfghjkl")
        bottom = set("zxcvbnm")

        results: List[str] = []

        for word in words:
            as_set = set(word.lower())
            if as_set.issubset(top) or as_set.issubset(middle) or as_set.issubset(bottom):
                results.append(word)

        return results
    
class Tests(unittest.TestCase):
    def test_example1(self):
        solution = Solution()
        self.assertEqual(solution.findWords(["Hello","Alaska","Dad","Peace"]), ["Alaska","Dad"])

    def test_example2(self):
        solution = Solution()
        self.assertEqual(solution.findWords(["omk"]), [])

    def test_example3(self):
        solution = Solution()
        self.assertEqual(solution.findWords(["adsdf","sfd"]), ["adsdf","sfd"])

if __name__ == '__main__':
    unittest.main()

File: elixir/500-KeyboardRow/main.exs
defmodule Solution do
  @spec find_words(words :: [String.t]) :: [String.t]
  def find_words(words) do
    top = "qwertyuiop" |> String.graphemes |> MapSet.new
    middle = "asdfghjkl" |> String.graphemes |> MapSet.new
    bottom = "zxcvbnm" |> String.graphemes |> MapSet.new

    check_word = fn(word) ->
      as_set = word |> String.downcase 
                    |> String.graphemes 
                    |> MapSet.new

        MapSet.subset?(as_set, top) or MapSet.subset?(as_set, middle) or MapSet.subset?(as_set, bottom)
    end

    for word <- words, check_word.(word), do: word
  end
end

IO.inspect(Solution.find_words(["Hello","Alaska","Dad","Peace"]))
IO.inspect(Solution.find_words(["omk"]))
IO.inspect(Solution.find_words(["adsdf","sfd"]))